Glshaders in ronin

XiNNiW opened the pull request:
Hi @neauoire,

I took a shot at cleaning up the gl tools and packaging them up in their own module.

I also made a library function called `kaleid` that makes a kaleidoscope of the current image on the canvas. It was helpful to have a concrete example to build toward.

This was fun to play with and I'm happy to iterate on this more!

I'd love to hear what you think.

Thanks,

David

neauoire commented:
This is very nice, almost perfect ;)
I was wondering for the kaleid shader, I don't think the way to go is to add new changes to the source everytime you want to add a new shader, the idea would be to do this in lisp space.

The simplest way would be to be able to load shaders, like we're currently loading an image, so it's kept in the Source module(source.js), and linked into the lisp file. Do you want to try and make it so you can load shaders instead of storing kaleid as a js function?

XiNNiW commented:
Thanks! That is a really cool idea. The user could load in a shader and
then use it however they want. I wonder how arguments could be bound to the
shader (for example the number of sides in the kaleidescope).

 I'll take another shot at it. Maybe the user could reference a json file
on their machine that contains the argument types and the shader code


On Thu, Apr 2, 2020, 5:56 PM Devine Lu Linvega <notifications@github.com>
wrote:

> This is very nice, almost perfect ;)
> I was wondering for the kaleid shader, I don't think the way to go is to
> add new changes to the source everytime you want to add a new shader, the
> idea would be to do this in lisp space.
>
> The simplest way would be to be able to load shaders, like we're currently
> loading an image, so it's kept in the Source module(source.js), and linked
> into the lisp file. Do you want to try and make it so you can load shaders
> instead of storing kaleid as a js function?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hundredrabbits/Ronin/pull/122#issuecomment-608111624>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEYDSX3BV2TBGPN3IXF3I3RKUCXDANCNFSM4L23HGZQ>
> .
>

XiNNiW commented:
What if it worked like this? 
There is a library function called `glsl` that you could use like so:
```
(import "path/to/image")
(glsl "path/to/kaleid.shader.json" 4 (rect 0 0 600 600) )
```
Where the "kaleid.shader.json" was a file defining the shader code and the meta-data required to tell Ronin how to bind its arguments and attributes. 

Here is a [gist](https://gist.github.com/XiNNiW/a19813f4042085963f110065e2832b6d) describing how the shader.json might look.

I think this solves the problem of not having to change the source code of Ronin every time you want a new shader. However, there are things I don't love about this approach:
1. Representing glsl code in json is a bit clunky
1. The design might be assuming that the user knows how Ronin works on the inside.
1. The json file is big. Although, this could be partially addressed by making the vertex shader optional and making it bind the canvas to a texture and the frame to position by default

Anyway, I would love to hear what you think. :)

XiNNiW commented:
I took a shot at implementing what I described above. It seemed to work pretty good! I updated the `kaleid.lisp` example in the example folder to match the new implementation.

I had to add some functions outside the gl module in order to make it work though. The most challenging aspect is correctly binding the attributes that the user specifies. These attributes could represent colors, shapes, or other arbitrary data. I implemented a method on `rect` called `getGlRepresentation` that can return the shape in a format acceptable to the gl shader attribute. It feels like each data type (most common being colors and shapes) would need to be able to tell you their glRepresentation in order to allow binding as an attribute. Is that OK?

There are probably other ways to solve this problem that I'm not seeing. 

There are also some improvements on this that would be easy to make, such as adding a default vertex shader that binds the canvas and the frame.

neauoire commented:
I'm a bit swamped at the moment, give me a day or two to look this over :)
Sorry for the delay ^^;

XiNNiW commented:
No problem and no rush. Sorry I kept spamming the thread. I think I may have problem sniped myself with this one. :)

I hope your well and staying safe.
